src/background_music/Integrator.h

```diff
@@ -0,0 +1,165 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions, and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions, and the following disclaimer in the documentation
+ *    provided with the distribution.
+ *
+ * This software is provided "as is" and without any warranty, express or
+ * implied. In no event shall the authors be liable for any damages arising from
+ * the use of this software.
+ */
+
+#include <algorithm>
+#include <cmath>
+#include <type_traits>
+
+/**
+ * Implements a classic RC-filter using an recursive filter with one input and
+ * one feedback coefficient.
+ * The impulse response of the filter is a decaying exponent that will reach
+ * \em 1/e of its original value after the number of "characteristic" samples.
+ * Characteristic samples larger than about 400M may yield inaccurate results
+ * due to repeated calculations and floating point precision.
+ */
+class IntegratorCoefficients
+{
+    /*
+     * The constructed defaults constitute an integrator that has zero
+     * characteristic samples, meaning it just echoes the input.
+     */
+    double history_multiplier_ = 0;
+    double input_multiplier_ = 1;
+
+public:
+    constexpr IntegratorCoefficients() = default;
+
+    void set_samples(const double characteristic_samples,
+                     const double scale = 1.0)
+    {
+        const double count = fabs(characteristic_samples);
+        history_multiplier_ = count > 0 ? exp(-1.0 / count) : 0.0;
+        set_scale(scale);
+    }
+
+    void set_seconds_for_rate(const double seconds, const int rate,
+                              const double scale = 1.0)
+    {
+        set_samples(seconds * static_cast<double>(rate), scale);
+    }
+
+    void set_scale(const double scale)
+    {
+        input_multiplier_ = fabs(scale) * (1.0 - history_multiplier_);
+    }
+
+    /**
+     * Integrates value input using history as integrated value.
+     * @param history The historic integrated value
+     * @param input The new input
+     */
+    void integrate(double & history, const double input) const
+    {
+        history = history_multiplier_ * history + input_multiplier_ * input;
+    }
+};
+
+class Integrator : public IntegratorCoefficients
+{
+    double integrated_ = 0;
+
+public:
+    double integrate(const double input)
+    {
+        IntegratorCoefficients::integrate(integrated_, input);
+        return integrated_;
+    }
+
+    template<typename T>
+    T integrate(const T input)
+    {
+        return static_cast<T>(integrate(static_cast<double>(input)));
+    }
+
+    void set_output(const double new_value) { integrated_ = new_value; }
+};
+
+/**
+ * Implements an envelope function that follows rises in input - attack -
+ * immediately and on input that is lowe than the current output, optionally the
```

Comment:
Typo: lowe(r)

Reply:
Fixed as part of [this commit](https://github.com/audacious-media-player/audacious-plugins/commit/ff69402b28597d41ac73c29e04ec9dbfd79d5a8f)